### PR TITLE
Fix amp autocast 

### LINKED ISSUE #6077: Mixed precision not working with v 1.2
## 🐛 Bug

After updating to 1.2 from 1.1.1, automatic mixed precision stopped working. Everything's float32 and getting CUDA OOM when I shouldn't get it (with float16 tensors). Worked fine on 1.1.1.

Here's my Trainer args (maybe there's a conflicting combo of args or something):

```
Trainer(logger=logger,
        callbacks=[checkpoint_callback, lr_monitor],
        default_root_dir=None,
        gradient_clip_val=args.gradient_clip_val,
        gpus=args.gpus,
        auto_select_gpus=False,
        log_gpu_memory=None,
        progress_bar_refresh_rate=1,
        check_val_every_n_epoch=args.check_val_every_n_epoch,
        overfit_batches=0.,
        fast_dev_run=False,
        accumulate_grad_batches=1,
        max_epochs=args.max_epochs,
        limit_train_batches=vars(args).get('limit_train_batches', 1.),
        val_check_interval=args.val_check_interval,
        limit_val_batches=args.limit_val_batches,
        accelerator='ddp',
        sync_batchnorm=True,
        precision=args.precision,
        weights_summary='top',
        weights_save_path=None,
        num_sanity_val_steps=args.num_sanity_val_steps,
        resume_from_checkpoint=None,
        benchmark=False,
        deterministic=False,
        reload_dataloaders_every_epoch=True,
        terminate_on_nan=False,
        prepare_data_per_node=True,
        amp_backend='native',
        profiler=args.profiler)
```

### Environment

- GCP VM with V100 GPU(s)
- NVIDIA-SMI 450.51.06    Driver Version: 450.51.06    CUDA Version: 11.0

Don't really have the time to go deeper than this, but just rolled back to 1.1.1 and everything's fine.

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #6077

We had an incorrect yield logic for the context manager. 
Modifications to the tests don't pass on master, validating the fix.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/models/test_amp.py [hunk 1]
```diff
@@ -63,7 +63,7 @@ def test_amp_single_gpu_ddp_spawn(tmpdir):
     model = BoringModel()
     # tutils.run_model_test(trainer_options, model)
     trainer.fit(model)
-
+    assert torch.is_autocast_enabled()
```
Comment (SeanNaren) [inline tests/models/test_amp.py:66]: huh, is this true? I thought it will only be enabled in the context region (training step/val step test step)

Comment (awaelchli) [reply tests/models/test_amp.py]: yes you're right

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]
